reps/2024-07-01-raycluster-crd-status-improvement.md

```diff
@@ -0,0 +1,137 @@
+
+4. Add a new condition `Status.Conditions[HeadReady]` to indicate whether the Ray head Pod is ready for requests or not. This is also useful for RayJob and RayService, where KubeRay needs to communicate with the Ray head Pod.
+
+5. Redefine `rayv1.Ready`.
+    * As mentioned in the “Context” section, when all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready` ([code](https://github.com/ray-project/kuberay/blob/8453f9b6c749abf3b72b09e0d41032bbfcf17367/ray-operator/apis/ray/v1/raycluster_types.go#L113)). It is highly possible that the RayCluster will remain in the `ready` state even if some incidents occur, such as manually deleting the head Pod.
+    * Here, we redefine `ready` as follows:
+        * When all Pods in the RayCluster are running for the first time, the RayCluster CR’s `Status.State` will transition to `ready`.
+        * After the RayCluster CR transitions to `ready`, KubeRay only checks whether the Ray head Pod is ready to determine if the RayCluster is `ready`.
+        * This new definition is mainly to maintain backward compatibility.
+    * TODO
+        * I am considering making `ready` a condition in `Status.Conditions` and removing `rayv1.Ready` from `Status.State`. Kubernetes Pod also makes a [PodCondition](https://pkg.go.dev/k8s.io/api/core/v1#PodConditionType), but this may be a bit aggressive.
+        * In the future, we can expose a way for users to self-define the definition of `ready` ([#1631](https://github.com/ray-project/kuberay/issues/1631)).
+
+## Implementation
+
+### Step 1: Introduce Status.Conditions
+
+```go
+type RayClusterStatus struct {
+       ...
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type RayClusterConditionType string
+
+const (
+	HeadReady RayClusterConditionType = "HeadReady"
+	// Pod failures. See DeploymentReplicaFailure and
+// ReplicaSetReplicaFailure for more details.
+RayClusterReplicaFailure RayClusterConditionType = "ReplicaFailure"
+)
+```
+
+* Future works
+    * Add `RayClusterAvailable` or `RayClusterReady` to `RayClusterConditionType` in the future for the RayCluster readiness.
+        * Reference: [DeploymentAvailable](https://github.com/kubernetes/api/blob/857a946a225f212b64d42c68a7da0dc44837636f/apps/v1/types.go#L532-L542), [LeaderWorkerSetAvailable](https://github.com/kubernetes-sigs/lws/blob/557dfd8b14b8f94633309f6d7633a4929dcc10c3/api/leaderworkerset/v1/leaderworkerset_types.go#L272)
+    * Add `RayClusterK8sFailure` to surface the Kubernetes resource failures that are not Pods.
```

Comment:
It might actually be worthwhile to introduce a generic condition type like `RayClusterInitializing` that encapsulates all the other resource dependencies that aren't pods.

Reply:
Probably not? Additional conditions for other resource dependencies are good, but a `RayClusterInitializing` may be too vague and could be easily misused. It smells like another `Ready`.

Reply:
`RayClusterK8sFailure` is pretty vague too. Maybe `RayClusterResourceInitialization` is more specific?

Reply:
RayClusterK8sFailure is indeed vague. I think conditions RayClusterXXXFailure, where XXX is specific, are better.

Reply:
I am still considering whether to add the field or not. Maybe creating K8s events for the non-Pod failures is enough?